### PR TITLE
Tags for Asset and AssetRegistry

### PR DESCRIPTION
**Pull request by Maksims**
Ability to tag asset, and query asset registry by tags, allowing simple single tag searches, as well as more complicated AND/OR scenarios.

TagsCache is internal class used by AssetRegistry, and will be used by Scene to tag entities.
It takes benefit of provided `key` index, to build internal indexes for making find queries fast and efficient.

### CODE REVIEW

File: src/core/core_tags.js [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+            return items;
+        }
+    };
+
+
+
+    /**
+    * @name pc.Tags
+    * @class Set of tag names
+    * @constructor Create an instance of an Tags.
```
Comment (willeastcott) [inline src/core/core_tags.js]: an -> a

File: src/core/core_tags.js [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+    * @param {Object} parent (optional) Parent object who tags belong to.
+    * Note: Tags are used as addition of `pc.Entity` and `pc.Asset` as `tags` field.
```
Comment (willeastcott) [inline src/core/core_tags.js]: pc.Entity support would be cool, but it's not in this PR, right?

File: src/core/core_tags.js [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+    Tags.prototype = {
+        /**
+        * @function
+        * @name pc.Tags#add
+        * @description Add a tag, duplicates are ignored. Can be array or comma separated arguments for multiple tags.
+        * @param {String} name of a tag, or array of tags
+        * @returns {Boolean} true if any tag were added
+        * @example
+        * tags.add('level-1');
+        * @example
```
Comment (willeastcott) [inline src/core/core_tags.js]: Is this how the JSdoc example tag works? I thought you could only have one. Maybe I'm wrong - I never tried using multiple...

**Comment by willeastcott** [comment]
Looks good to me apart from the minor comments I had. :)

File: src/core/core_tags.js (hunk 2 above)
Comment (Maksims) [inline src/core/core_tags.js]: TagsCache and Tags is implemented to be agnostic regarding what it's tagging. So can be added later, although currently it is not part of this PR, as there are few questions of where TagsCache should sit on: Application or Scene, or Root node?
As currently it is: `app.root.findByName`, so makes sense to do: `app.root.findByTag`, although root is a `GraphNode`, so a bit of technical design is required here.

File: src/core/core_tags.js (hunk 3 above)
Comment (Maksims) [inline src/core/core_tags.js]: I've looked at places around, seemed like used practice for multiple examples.

File: src/core/core_tags.js [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    * @constructor Create an instance of a Tags.
+    * @param {Object} parent (optional) Parent object who tags belong to.
+    * Note: Tags are used as addition of `pc.Entity` and `pc.Asset` as `tags` field.
+    */
+
+    /**
+     * @event
+     * @name pc.Tags#add
+     * @param {String} tag Name of a tag added to a set.
+     * @param {Object} parent Parent object who tags belong to.
```
Comment (vkalpias) [inline src/core/core_tags.js]: I think this would be better: "Parent object to add the tags to"

File: src/core/core_tags.js [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+    */
+
+    /**
+     * @event
+     * @name pc.Tags#remove
+     * @param {String} tag Name of a tag removed from a set.
+     * @param {Object} parent Parent object who tags belong to.
```
Comment (vkalpias) [inline src/core/core_tags.js]: Similar to above comment

File: src/core/core_tags.js [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    */
+
+    /**
+     * @event
+     * @name pc.Tags#change
+     * @param {Object} parent Parent object who tags belong to.
```
Comment (vkalpias) [inline src/core/core_tags.js]: Similar to above comment

File: src/core/core_tags.js [hunk 7]
```diff
@@ -0,0 +1,410 @@
...
+     * @description fires when any change add/remove been made.
```
Comment (vkalpias) [inline src/core/core_tags.js]: Fires after any add / remove operation has been made

File: src/core/core_tags.js [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        }
+    };
+
+
+
+    /**
+    * @name pc.Tags
+    * @class Set of tag names
+    * @constructor Create an instance of a Tags.
+    * @param {Object} parent (optional) Parent object who tags belong to.
```
Comment (vkalpias) [inline src/core/core_tags.js]: If this is optional then surround the 'parent' with brackets like so [parent]

File: src/core/core_tags.js [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+        pc.events.attach(this);
+    };
+
+    Tags.prototype = {
+        /**
+        * @function
+        * @name pc.Tags#add
+        * @description Add a tag, duplicates are ignored. Can be array or comma separated arguments for multiple tags.
+        * @param {String} name of a tag, or array of tags
```
Comment (vkalpias) [inline src/core/core_tags.js]: should be: @param {String} name Name of a tag, or array of tags.

File: src/core/core_tags.js [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+
+            return changed;
+        },
+
+
+        /**
+        * @function
+        * @name pc.Tags#remove
+        * @description Remove tag.
+        * @param {String} name of a tag or array of tags
```
Comment (vkalpias) [inline src/core/core_tags.js]: Should be {String} name Name of a tag or array of tags

File: src/core/core_tags.js [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+            this.fire('change', this._parent);
+        },
+
+
+        /**
+        * @function
+        * @name pc.Tags#has
+        * @description Check if tags satisfy filters.
+        * Filters can be provided by simple name of tag, as well as by array of tags.
+        * When array is provided it will check if tags contain each tag within the array.
```
Comment (vkalpias) [inline src/core/core_tags.js]: When 'an' array is provided

File: src/core/core_tags.js [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        * If any of comma separated argument is satisfied, then it will return true.
```
Comment (vkalpias) [inline src/core/core_tags.js]: I think you can drop the 'of' here

**Pull request merged**